src/couch_replicator.erl

```diff
@@ -52,17 +52,32 @@ replicate(PostBody, Ctx) ->
     #rep{id = RepId, options = Options, user_ctx = UserCtx} = Rep,
     case get_value(cancel, Options, false) of
     true ->
-        case get_value(id, Options, nil) of
+        CancelRepId = case get_value(id, Options, nil) of
         nil ->
-            cancel_replication(RepId);
+            RepId;
         RepId2 ->
-            cancel_replication(RepId2, UserCtx)
+            RepId2
+        end,
+        case check_authorization(CancelRepId, UserCtx) of
+        authorized ->
+            cancel_replication(CancelRepId);
+        unauthorized ->
+            Msg = <<"Can't cancel a replication triggered by another user">>,
+            throw({unauthorized, Msg});
+        not_found ->
+            {error, not_found}
         end;
     false ->
-        {ok, Listener} = rep_result_listener(RepId),
-        Result = do_replication_loop(Rep),
-        couch_replicator_notifier:stop(Listener),
-        Result
+        case check_authorization(RepId, UserCtx) of
+        unauthorized ->
+            Msg = <<"Replication already started by another user">>,
+            throw({unauthorized, Msg});
```

Comment:
Another thought: If you use `couch_httpd: verify_is_server_admin/1`, you can simplify these case clauses. No need to throw anything here or above.